### PR TITLE
[curl] Add feature idn2

### LINKED ISSUE #14798: [curl:x64-linux] build failure
**Host Environment**
- OS: Arch Linux on WSL
- Compiler: revision

**To Reproduce**
`./vcpkg install cpr`

**Failure logs**
lot of patches here
```
Starting package 65/68: curl:x64-linux
Building package curl[core,non-http,openssl,ssl]:x64-linux...
-- Using cached /var/cache/vcpkg/curl-curl-315ee3fe75dade912b48a21ceec9ccda0230d937.tar.gz
-- Extracting source /var/cache/vcpkg/curl-curl-315ee3fe75dade912b48a21ceec9ccda0230d937.tar.gz
-- Applying patch 0002_fix_uwp.patch
-- Applying patch 0004_nghttp2_staticlib.patch
-- Applying patch 0005_remove_imp_suffix.patch
-- Applying patch 0006_fix_tool_depends.patch
-- Applying patch 0007_disable_tool_export_curl_target.patch
-- Applying patch 0009_fix_openssl_config.patch
-- Applying patch 0010_fix_othertests_cmake.patch
-- Applying patch 0011_fix_static_build.patch
-- Using source at /usr/share/vcpkg/buildtrees/curl/src/da0230d937-b280319101.clean
-- Configuring x64-linux-dbg
-- Configuring x64-linux-rel
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Installing: /usr/share/vcpkg/packages/curl_x64-linux/share/curl/curl-config
-- Fixing pkgconfig file: /usr/share/vcpkg/packages/curl_x64-linux/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_idn2_RELEASE:CHECK_LIB_idn2_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "idn2" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:185 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)


Error: Building package curl:x64-linux failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: curl:x64-linux
  Vcpkg version: 2020.11.12-unknownhash
```

**Additional context**
The command `idn2` works, and `vcpkg install idn2` says it doesn't find `unistring`.

Thank you

### LINKED ISSUE #14834: ./vcpkg install curl[tool] fails
-- Installing: /home/martti/temp/vcpkg/packages/curl_x64-linux/share/curl/curl-config
-- Fixing pkgconfig file: /home/martti/temp/vcpkg/packages/curl_x64-linux/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_lber_RELEASE:CHECK_LIB_lber_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "lber" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:185 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)

Error: Building package curl:x64-linux failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: curl:x64-linux
  Vcpkg version: 2020.11.12-unknownhash

Additionally, attach any relevant sections from the log files above.

### PR DESCRIPTION
**Pull request by JackBoosY**
- add curl feature idn2 to avoid to find the system idn2 library and headers.

Fixes #14798.

Note: already test all features on `x86-windows` and `x64-windows-static` successfully.

### CODE REVIEW

File: ports/curl/0012-fix-dependency-idn2.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+-check_library_exists_concat("idn2" idn2_lookup_ul HAVE_LIBIDN2)
++if (CMAKE_USE_IDN2)
++    find_package(libidn2 REQUIRED)
++    add_definitions(-DHAVE_IDN2_H=1)
++    add_definitions(-DHAVE_LIBIDN2=1)
++    include_directories(${LIBIDN2_INCLUDE_DIRS})
++    link_libraries(${LIBIDN2_LIBRARIES})
++else()
++    add_definitions(-DHAVE_IDN2_H=0)
++    add_definitions(-DHAVE_LIBIDN2=0)
```
Comment (JackBoosY) [inline ports/curl/0012-fix-dependency-idn2.patch:18]:
```suggestion
```

File: ports/curl/0012-fix-dependency-idn2.patch [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
++++ b/CMakeLists.txt
+@@ -621,7 +621,16 @@ if(NOT CURL_DISABLE_LDAPS)
+ endif()
+ 
+ # Check for idn
+-check_library_exists_concat("idn2" idn2_lookup_ul HAVE_LIBIDN2)
++if (CMAKE_USE_IDN2)
++    find_package(libidn2 REQUIRED)
++    add_definitions(-DHAVE_IDN2_H=1)
++    add_definitions(-DHAVE_LIBIDN2=1)
```
Comment (JackBoosY) [inline ports/curl/0012-fix-dependency-idn2.patch:13]:
```suggestion
+    add_definitions(-DHAVE_IDN2_H)
+    add_definitions(-DHAVE_LIBIDN2)
```

**Comment by klalumiere** [comment]
Hi JackBoosY!

I see that this pull request is still a work in progress. Do you have an ETA for when it's going to be ready? Right now the port `curl` and the 39 ports that depend on it are broken on Linux (I don't understand why the CI didn't catch this :slightly_frowning_face: ). If you tell me that your ETA is more a few days, I could do a 2-liner pull request that allows the port to work again on Linux while waiting for the cleaner solution you are developing, i.e. doing

`vcpkg_fixup_pkgconfig(SYSTEM_LIBRARIES pthread dl c)` -> `vcpkg_fixup_pkgconfig(SYSTEM_LIBRARIES pthread dl c lber ldap)`.
`
`Port-Version: 3` -> `Port-Version: 4`

is an ugly (or more precisely, an incorrect) but effective way to make `curl` work again. :slightly_smiling_face:

**Comment by JackBoosY** [comment]
Hi @klalumiere, it should be fixed on my latest commit, can you test it?
The reason for this issue is that some curl dependencies are not added as other dependencies (add `option` to `CMakeLists.txt` and use `find_package`/`find_library` to find them), and I've disabled them.

**Comment by JackBoosY** [comment]
Also, @martti-cloudkey can you please test this PR?

Thanks.

**Comment by klalumiere** [comment]
The problem with `lber` has been replaced by a problem with `ldap` :slightly_frowning_face: 

```shell
-- Installing: /home/klalumiere/Downloads/vcpkg/packages/curl_x64-linux/share/curl/curl-config
-- Fixing pkgconfig file: /home/klalumiere/Downloads/vcpkg/packages/curl_x64-linux/lib/pkgconfig/libcurl.pc
-- CHECK_LIB_ldap_RELEASE:CHECK_LIB_ldap_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:233 (message):
  Library "ldap" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:187 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)


Error: Building package curl:x64-linux failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: curl:x64-linux
  Vcpkg version: 2020.11.12-unknownhash

Additionally, attach any relevant sections from the log files above.
[21:04:48] klalumiere@WKS-000791:~/Downloads/vcpkg$ git status
On branch dev/jack/14798
Your branch is up to date with 'origin/dev/jack/14798'.

nothing to commit, working tree clean
```

**Comment by JackBoosY** [comment]
@KaibaLopez Can you just modify _VCPKG_ROOT/port/curl/portfile.cmake_ line 71:
```
...
    OPTIONS ${FEATURE_OPTIONS}
        -DCURL_DISABLE_LDAP=OFF
        ${UWP_OPTIONS}
...
```
And try again?

Thanks.

**Comment by JackBoosY** [comment]
@klalumiere  Any feedback with my [comment](https://github.com/microsoft/vcpkg/pull/14807#issuecomment-736178259)?

**Comment by klalumiere** [comment]
> @KaibaLopez Can you just modify _VCPKG_ROOT/port/curl/portfile.cmake_ line 71:
> 
> ```
> ...
>     OPTIONS ${FEATURE_OPTIONS}
>         -DCURL_DISABLE_LDAP=OFF
>         ${UWP_OPTIONS}
> ...
> ```
> 
> And try again (do not use my changes)?
> 
> Thanks.

I had exactly the same problem with `-DCURL_DISABLE_LDAP=OFF` (that is, `./vcpkg install curl` fails). _However_, adding `-DCURL_DISABLE_LDAP=ON` _did_ solve the problem (I think there was a confusion with the double negative :slightly_smiling_face:).

**Comment by JackBoosY** [comment]
@klalumiere Yep, my mistake. It should be `ON`. I will continue fix this issue today.

**Comment by JackBoosY** [comment]
Need restore my changes about lber2.

File: ports/curl/portfile.cmake [hunk 3]
```diff
@@ -57,6 +59,7 @@ if(VCPKG_TARGET_IS_UWP)
     set(UWP_OPTIONS
         -DUSE_WIN32_LDAP=OFF
         -DCURL_DISABLE_TELNET=ON
+        -DCURL_DISABLE_LDAP=ON
```
Comment (JackBoosY) [inline ports/curl/portfile.cmake:62]: This should not be simply disabled.

**Comment by JackBoosY** [comment]
@klalumiere Can you remove `DCURL_DISABLE_LDAP=ON`, install `libldap2-dev` and `libsasl2-dev` then try again?

Thanks.

**Comment by klalumiere** [comment]
So as you saw on #14865, I'm not able to reproduce the problem I had anymore on master. However, I still have problem with the branch `JackBoosY:dev/jack/14798`. I built in a Docker container running the image `rikorose/gcc-cmake`. I installed `vim` with

```shell
apt-get update && apt-get install vim --no-install-recommends
```

and that's all. Without any changes to the branch, I ran `./vcpkg install curl:x64-linux-dynamic` with the triplet [`x64-linux-dynamic` from the doc](https://vcpkg.readthedocs.io/en/latest/examples/overlay-triplets-linux-dynamic/), i.e.

```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic) # This changed from static to dynamic

set(VCPKG_CMAKE_SYSTEM_NAME Linux)
```

It fails, with the output

```shell
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:235 (message):
  Unhandled string
  "/root/vcpkg/packages/curl_x64-linux-dynamic/lib/pkgconfig/../../lib/libssl.so
  /root/vcpkg/packages/curl_x64-linux-dynamic/lib/pkgconfig/../../lib/libcrypto.so
  /root/vcpkg/packages/curl_x64-linux-dynamic/lib/pkgconfig/../../lib/libz.so"
  was found! If it is a system library use the SYSTEM_LIBRARIES parameter for
  the vcpkg_fixup_pkgconfig call! Otherwise, correct the *.pc file or add the
  case to vcpkg_fixup_pkgconfig
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:299 (vcpkg_fixup_pkgconfig_check_files)
  ports/curl/portfile.cmake:188 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:136 (include)


Error: Building package curl:x64-linux-dynamic failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: curl:x64-linux-dynamic
  Vcpkg version: 2020.11.12-unknownhash

Additionally, attach any relevant sections from the log files above.
```

I'll try removing `-DCURL_DISABLE_LDAP=ON` next, although it doesn't seems to be related to the error. I'll try installing the libs you suggested after. It might take a while. Building is super slow in a container :slightly_frowning_face:

**Comment by klalumiere** [comment]
Actually... Shouldn't the string `/root/vcpkg/packages/curl_x64-linux-dynamic/lib/pkgconfig/../../lib/libssl.so` be replaced by `-lssl` in the pkgconfig? This seems to be what is done elsewhere in the _portfile.cmake_. Oh, I see, e.g. _portfile.cmake:139_

```cmake
vcpkg_replace_string(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc "${CURRENT_INSTALLED_DIR}/debug/lib/libssl.a" "-lssl")
```
There should also be 

```cmake
vcpkg_replace_string(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc "${CURRENT_INSTALLED_DIR}/debug/lib/libssl.so" "-lssl")
```
Couldn't we use a regular expression instead? Something in the spirit of

```cmake
vcpkg_replace_string(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc "${CURRENT_INSTALLED_DIR}/debug/lib/lib(.*)\\.so" "-l\1")
```

**Comment by klalumiere** [comment]
We could add a function

```cmake
function(vcpkg_replace_string_regex filename match_string replace_string)
    file(READ ${filename} _contents)
    string(REGEX REPLACE "${match_string}" "${replace_string}" _contents "${_contents}")
    file(WRITE ${filename} "${_contents}")
endfunction()
```
and then use it as I was suggesting, but without my mistake :laughing: 
```cmake
vcpkg_replace_string_regex(${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/libcurl.pc
    "${CURRENT_INSTALLED_DIR}/debug/lib/lib(.*)\.so" "-l\\1")

**Comment by klalumiere** [comment]
I had practically the same error than [here](https://github.com/microsoft/vcpkg/pull/14807#issuecomment-741424683) without `-DCURL_DISABLE_LDAP=ON`, the only difference being of course the line number, i.e. `ports/curl/portfile.cmake:187 (vcpkg_fixup_pkgconfig)

**Comment by JackBoosY** [comment]
cc @ras0219 for review this PR. I'm not sure my changes is correct.

**Comment by klalumiere** [comment]
Merging `master` did fix the build on Linux with the triplet [`x64-linux-dynamic`](https://vcpkg.readthedocs.io/en/latest/examples/overlay-triplets-linux-dynamic/). :partying_face: 

The systematic warning when we build on a system different than Windows is a bit unfortunate. :slightly_frowning_face: . Could this warning instead be added to cURL's `CMakeLists.txt`? I believe that once inside the CMakeList.txt, we might have the information necessary to know if `libldap2-dev` and `libsasl2-dev` will be required for the build. We could print the warning only if we detect that they are necessary, and that they can't be found. :thinking:

**Comment by JackBoosY** [comment]
@klalumiere Adding the message to `CMakeLists.txt` is not a good choice, because I'm not sure if these two packages are needed, and curl builds successfully in my ubuntu without that.

**Comment by klalumiere** [comment]
> @klalumiere Adding the message to `CMakeLists.txt` is not a good choice, because I'm not sure if these two packages are needed, and curl builds successfully in my ubuntu without that.

I'm not sure I understand. Right now the warning is inside the `portfile.cmake`. How is it better to put the warning inside the `portfile.cmake` than to put it in the `CMakeLists.txt`? As I mentioned above, the advantages of putting it inside the `CMakeLists.txt` is that you can search (`find_library` or `find_package`) for e.g. `ldap2` and, if it is not found, print the warning. What are the advantages of putting the warning inside te `portfile.cmake`?

It also works on my machine (Ubuntu 20.04). `dpkg --list libldap2-dev` tells me that this package is installed, but `dpkg --list libsasl2-dev` informs me that there is no such package on my machine. It probably would be worth it to verify what happen inside a Docker container where none of the packages are installed. If none are required, we can juste remove the warning altogether.

**Comment by JackBoosY** [comment]
@klalumiere Normally, the warning message like that always put into `portfile.cmake`, such as [pixel](https://github.com/microsoft/vcpkg/blob/d0531aac1f44cdc1ae6987e7b0b4f60c72304994/ports/pixel/portfile.cmake), [glfw3](https://github.com/microsoft/vcpkg/blob/6d1cd558e6183e1b821150681b0c703badcdb591/ports/glfw3/portfile.cmake), [spdk](https://github.com/microsoft/vcpkg/blob/09a647a5261e981cbbc8219292c838490ed0b06f/ports/spdk/portfile.cmake) etc.
We need to detect and prepare these dependencies before configuration, rather than dynamically checking during configuration. Otherwise vcpkg will report error logs instead of error message.

**Comment by klalumiere** [comment]
I retried to install `curl` inside the Docker image `rikorose/gcc-cmake` with your latest changes. I tried with the triplets `x64-linux` _and_ `x64-linux-dynamic`. Both work, and I made sure that neither `libldap2-dev` nor `libsasl2-dev` are installed. Hence, I don't think that the warning is required: it works without `libldap2-dev` and `libsasl2-dev`.

Good point about the error logs instead of error message. I think the way [spdk](https://github.com/microsoft/vcpkg/blob/09a647a5261e981cbbc8219292c838490ed0b06f/ports/spdk/portfile.cmake) does it is smart, i.e. it uses `FIND_PATH` to print (an error in this case) only if there's actually a problem.

File: ports/curl/portfile.cmake [hunk 4]
```diff
@@ -28,6 +29,10 @@ if("sectransp" IN_LIST FEATURES AND NOT VCPKG_TARGET_IS_OSX)
     message(FATAL_ERROR "sectransp is not supported on non-Apple platforms")
 endif()
 
+if (NOT VCPKG_HOST_IS_WINDOWS)
+    message(WARNING "Feature ldap and ldaps may need to use the system package manager to install libldap2-dev and libsasl2-dev.")
+endif()
+
```
Comment (JackBoosY) [inline ports/curl/portfile.cmake:35]:
```suggestion
```

**Comment by BillyONeal** [comment]
@JackBoosY Why is this "requires:discussion"? It looks like a normal bugfix to me?

**Review by BillyONeal — APPROVED** [review]

**Comment by JackBoosY** [comment]
@BillyONeal Discussion for feature `lber2` in curl's source. I'm not sure which method we should choose to solve that.

**Review by ras0219-msft — APPROVED** [review]

**Pull request merged**